openfe/protocols/openmm_utils/utils.py

```diff
@@ -0,0 +1,105 @@
+# This code is part of OpenFE and is licensed under the MIT license.
+# For details, see https://github.com/OpenFreeEnergy/openfe
+"""
+Reusable unclassified utility methods for OpenMM-based alchemical
+Protocols.
+"""
+from pathlib import Path
+from numpy import typing as npt
+from openmm import app
+
+
+def subsample_omm_topology(
+    topology: app.Topology,
+    subsample_indices: tuple,
+) -> app.Topology:
+    """
+    Taking an initial app.Topology and a tuple of atom indices, generate
+    a new app.Topology which only contains those atoms and any bonds which
+    are involved between these atoms.
+
+    Parameters
+    ----------
+    topology : app.Topology
+      Initial Topology to subsample from
+    subsample_indices : tuple
+      Tuple of atom indices for the atoms we want to subsample from the
+      input Topology.
+
+    Returns
+    -------
+    top : app.Topology
+      Subsampled Topology.
+    """
+    top = app.Topology()
+    chains = set()
+    residues = set()
+    old_to_new_atom_map = {}
+
+    # First pass - get chain & resid info
+    for at in topology.atoms():
+        if at.index in subsample_indices:
+            residues.add(at.residue.index)
+            chains.add(at.residue.chain.index)
+
+    # Now let's actually add things
+    for chain_id, chain in enumerate(topology.chains()):
+        if chain.index in chains:
```

Comment:
if you instead do `if not chain.index in chains: continue` you can avoid going 6 layers deep on the indentation and running out of known universe